Getting help guide (#2828)
NomicFoundation/hardhat: docs/src/content/hardhat-runner/docs/guides/getting-help.md

# Getting help

Hardhat has a strong community of users willing to help you in times of trouble. Please, read this guide entirely to learn where and how to ask for help more effectively.

The first place to look for answers is the [GitHub Discussions section in the Hardhat repository.](https://github.com/nomiclabs/hardhat) We recommend you search there first, as the answer may already exist.

> If you can't find what you are looking for on GitHub Discussions, you can create a new Discussion.

Reviewer: ```suggestion
If you can't find what you are looking for on GitHub Discussions, you can [create a new Discussion](https://github.com/NomicFoundation/hardhat/discussions/new).
```

> If you weren't lucky on GitHub, or you prefer a real-time chat, you can join our [Discord Server](https://hardhat.org/discord). Please, read its rules, and ask for help in the right channel.

Reviewer 2: ```suggestion
If you didn't have any luck on GitHub, or if you prefer a real-time chat, you can join our [Discord Server](https://hardhat.org/discord). Please read its rules and ask for help in the right channel.
```

## Asking an effective question

> To get an answer quickly on GitHub Discussions and Discord, you need to make sure you write a good question first. To do so, you need to include:

Reviewer: I would say something like:

```suggestion
To increase the chances of getting an answer quickly on GitHub Discussions and Discord, you need to make sure you write a good question first. To do so, you need to include:
```

Feel free to ignore this though.

1. A clear description of what you are trying to do.
2. The results you are getting, and how they differ from what you expect.
3. Which version of Hardhat you are running, and which plugins and their versions you are using.

> 4. If possible, a link to a public repository to reproduce your problem.

Reviewer: ```suggestion
4. If possible, a link to a public repository to reproduce your problem or, even better, a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).
```

Reviewer 2: ```suggestion
4. Very specific and concise instructions on how to reproduce your problem. Ideally, provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Another good option is to provide a link to a public repository that provides an easy environment to reproduce the problem.
```

## Reporting a bug

> If you think you found a bug on Hardhat, and you didn't get help by following this guide, we'd appreciate if you [report it](https://github.com/NomicFoundation/hardhat/issues).

Reviewer 2: ```suggestion
If you think you've found a bug in Hardhat, please [report it](https://github.com/NomicFoundation/hardhat/issues).
```

> Please, follow these steps before opening a new issue:

Reviewer 2: ```suggestion
However, _before_ reporting a bug, please follow these steps to ensure that a new bug report is actually warranted:
```

1. Make sure you are using the latest version of Hardhat and its plugins. Your problem may already be fixed.

> 2. Try to identify if the issue comes from a plugin by running your project with some of them disabled. We only accept issues about plugins starting with `@nomiclabs/` and `@nomicfoundation/`.

Reviewer 2: ```suggestion
2. Try to determine whether the issue is coming from a plugin by running your project with some of them disabled. We only accept bug reports for plugins published by us, all of which start with `@nomiclabs/` and `@nomicfoundation/`. If you find an issue with a plugin published by someone else, you may be able to raise it with that publisher.
```

> 3. Use the search on GitHub to try to find other reports of the same problem. If you find one, please comment on the existing issue instead of creating a new one.

Reviewer: ```suggestion
3. Use the search on GitHub to try to find other reports of the same problem. If you find one and you can add useful new information, please comment on the existing issue instead of creating a new one.
```

Reviewer 2: @fvictorio I respectfully disagree with this suggested change. It's useful for us to know that more people are encountering the problem, even if those additional people don't have anything constructive to add.

Reviewer: That's fair, I just don't want to encourage "+1" comments. I thought about suggesting reacting with a :+1:, but I couldn't find a link to some canonical explanation on upvoting issues with reactions.

Author: Unfortunately `+1` is useful. Annoying, but useful. I'd blame GitHub and not the users for that.

If you followed the above steps, and you still need to create a new issue or comment on an existing one, please include all the information described in [Asking an effective question](#asking-an-effective-question).

> Finally, while submitting a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) is not strictly required, consider doing it, as it greatly increases the chances of your problem getting fixed quickly.

Reviewer: I like this link better:

```suggestion
Finally, while submitting a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) is not strictly required, consider doing it, as it greatly increases the chances of your problem getting fixed quickly.
```

> If you followed the above steps, and you still need to create a new issue or comment on an existing one, please include all the information described in [Asking an effective question](#asking-an-effective-question).
> 
> Finally, while submitting a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) is not strictly required, consider doing it, as it greatly increases the chances of your problem getting fixed quickly.

Reviewer 2: ```suggestion
Whenever reporting a bug, and ideally whenever commenting on an existing Issue, please include all the information described in [Asking an effective question](#asking-an-effective-question). By providing as much information as possible, you greatly increase the chances of your problem getting fixed quickly.
```
